lib/fizzy/instantiate.cpp

```diff
@@ -420,18 +420,29 @@ std::optional<FuncIdx> find_exported_function(const Module& module, std::string_
     return find_export(module, ExternalKind::Function, name);
 }
 
+ExecuteFunction::ExecuteFunction(Instance& instance, FuncIdx func_idx)
+  : m_function{[](void* context, Instance&, const Value* args, int depth) {
+        auto [func_idx, instance] = *static_cast<std::pair<FuncIdx, Instance*>*>(context);
+        return execute(*instance, func_idx, args, depth);
+    }},
+    m_context{new std::pair<FuncIdx, Instance*>(func_idx, &instance),
```

Comment:
I don't have full picture but this does not look good.

I want to thing about the "ExecuteFunction" as a variant of `(this instance, func_idx)`, `(other instance, func_idx)`, `(host function pointer, host function context pointer)`. So you need to store two pointers at most + enum to know the kind. You can start with `std::variant<WasmFunction, HostFunction>` as a start, but it might be difficult to convert it to C.

Reply:
( `(this instance, func_idx)` is not needed, because this now is used only for exported/imported functions)

Yeah, on the C side this is going to be a `union`, which might be ugly, but maybe we could add some helpers to simplify work with it... I'll think about it.

Anyway C API is the point of most complications with these refactorings...

Reply:
This should be union on C++ side too, the `std::variant` is just a wrapper for union. So you can use the C type (and add convenient wrapping by inheritance) if you expect `reinterpret_cast` is convenient.

The version with union can be like this:
```
struct F {
  Instance* instance;
  union {
    FuncIdx func_idx;
    HostFn host_fn;
  }
  void* host_context;
}
```
When `instance != nullptr` you use `func_idx`, otherwise this is a host function and you use `host_fn`.

Dropping `union{}` for now will also work, just the size is +33%.

Reply:
Main complications are around converting from C function pointer to C++ function pointer (different signature) and back.

E.g. when providing host imported function from C side, both context and C function pointer have to be saved somewhere, function pointer can't be just casted. (perhaps having a second `void*` in `ExecuteFunction` might solve it)
When returning re-exported C++ host function to C side, orignal context and original C++ host function pointer and also arrays with input parameter types have to be stored somewhere (perhaps having a second `void*` context in `FizzyExternalFunction` might solve it; parameter types should just point to original C++ module, that's another refactoring....)